metals/src/main/scala/scala/meta/metals/providers/DebugPayloadProvider.scala

```diff
@@ -0,0 +1,94 @@
+package scala.meta.metals.providers
+
+import java.util.zip.ZipEntry
+import java.util.zip.ZipOutputStream
+import java.io.ByteArrayOutputStream
+import java.nio.file.Files
+import java.time.Instant
+import scala.meta.AbsolutePath
+import com.typesafe.scalalogging.LazyLogging
+import io.circe.syntax._
+import io.circe.Json
+import cats.syntax.either._
+import scala.meta.metals.Uri
+import scala.meta.metals.Buffers
+import scala.meta.metals.Configuration
+import scala.meta.metals.search.DocumentIndex
+import scala.meta.metals.compiler.ScalacProvider
+import scala.util.control.NonFatal
+import org.langmeta.jsonrpc.Response
+
+class DebugPayloadProvider(
+    cwd: AbsolutePath,
+    latestConfig: () => Configuration,
+    scalacProvider: ScalacProvider,
+    buffers: Buffers,
+    documentIndex: DocumentIndex
+) extends LazyLogging {
+
+  implicit class XtensionByteArray(val s: String) {
+    def toByteArray: Array[Byte] = s.toCharArray.map(_.toByte)
+  }
+
+  private def zipWithFiles(f: Map[String, Array[Byte]]): Array[Byte] = {
+    val out = new ByteArrayOutputStream
+    val zip = new ZipOutputStream(out)
+    f.foreach {
+      case (name, bytes) =>
+        zip.putNextEntry(new ZipEntry(name))
+        zip.write(bytes)
+        zip.closeEntry()
+    }
+    zip.close()
+    out.toByteArray
+  }
+
+  private def currentUri(arguments: Option[Seq[Json]]): Option[Uri] =
+    arguments.flatMap(_.headOption).flatMap(_.asString).map(Uri.apply)
+
+  def generatePayload(
+      arguments: Option[Seq[Json]]
+  ): Either[Response.Error, Array[Byte]] =
+    Either.fromOption(
+      currentUri(arguments).map { uri =>
+        val logFile = cwd.resolve(".metals").resolve("metals.log").readAllBytes
+        val config = latestConfig().asJson.spaces2.toByteArray
+        val buildInfoEntry = scalacProvider
+          .configBySourceDirectory(uri)
+          .map { buildInfo =>
+            buildInfo.origin
+              .toRelative(cwd.resolve(".metals"))
+              .toString -> buildInfo.origin.readAllBytes
+          }
+          .toMap
+
+        val currentBuffer = buffers.read(uri).toByteArray
+        val semanticdbDocument = documentIndex
+          .getDocument(uri)
+          .map { doc =>
+            uri.toPath.getFileName.toString + ".semanticdb" -> doc.toString.toByteArray
```

Comment:
@olafurpg is there a way to get a nicer output for `m.Document` other than `.toString`?

Reply:
`doc.toByteArray` preserves the protobuf payload.

We should be very clear when the user generates the file that the source code for the open buffer is included and it's fine to remove the file leave if sharing code is a problem.